data/src/main/java/org/akvo/flow/data/repository/SurveyDataRepository.java

```diff
@@ -0,0 +1,192 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Akvo Flow.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package org.akvo.flow.data.repository;
+
+import android.database.Cursor;
+import android.text.TextUtils;
+
+import org.akvo.flow.data.datasource.DataSourceFactory;
+import org.akvo.flow.data.entity.ApiDataPoint;
+import org.akvo.flow.data.entity.ApiLocaleResult;
+import org.akvo.flow.data.entity.DataPointMapper;
+import org.akvo.flow.data.entity.SyncedTimeMapper;
+import org.akvo.flow.data.net.FlowRestApi;
+import org.akvo.flow.domain.entity.DataPoint;
+import org.akvo.flow.domain.exception.AssignmentRequiredException;
+import org.akvo.flow.domain.repository.SurveyRepository;
+
+import java.net.HttpURLConnection;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+import javax.inject.Inject;
+
+import retrofit2.HttpException;
+import rx.Observable;
+import rx.functions.Func1;
+
+public class SurveyDataRepository implements SurveyRepository {
+
+    private final DataSourceFactory dataSourceFactory;
+    private final DataPointMapper dataPointMapper;
+    private final SyncedTimeMapper syncedTimeMapper;
+    private final FlowRestApi restApi;
+
+    @Inject
+    public SurveyDataRepository(DataSourceFactory dataSourceFactory,
+            DataPointMapper dataPointMapper, SyncedTimeMapper syncedTimeMapper,
+            FlowRestApi restApi) {
+        this.dataSourceFactory = dataSourceFactory;
+        this.dataPointMapper = dataPointMapper;
+        this.syncedTimeMapper = syncedTimeMapper;
+        this.restApi = restApi;
+    }
+
+    @Override
+    public Observable<List<DataPoint>> getDataPoints(Long surveyGroupId,
+            Double latitude, Double longitude, Integer orderBy) {
+        return dataSourceFactory.getDataBaseDataSource()
+                .getDataPoints(surveyGroupId, latitude, longitude, orderBy).concatMap(
+                        new Func1<Cursor, Observable<List<DataPoint>>>() {
+                            @Override
+                            public Observable<List<DataPoint>> call(Cursor cursor) {
+                                return Observable.just(dataPointMapper.getDataPoints(cursor));
+                            }
+                        });
+    }
+
+    @Override
+    public Observable<Integer> syncRemoteDataPoints(final long surveyGroupId) {
+        return getServerBaseUrl()
+                .concatMap(new Func1<String, Observable<Integer>>() {
+                    @Override
+                    public Observable<Integer> call(final String serverBaseUrl) {
+                        return dataSourceFactory.getPropertiesDataSource().getApiKey()
+                                .concatMap(new Func1<String, Observable<Integer>>() {
+                                    @Override
+                                    public Observable<Integer> call(String apiKey) {
```

Comment:
I don't know yet what is a better way or even if there is one, but the nesting here feels strange for me in terms of code formatting.  If you don't have any idea right away I will merge it but would like to revisit it before the final merge into develop